### ISSUE TITLE
Fix tf.gather

### ISSUE THREAD

**Issue by SunYanCN** [andreyzharkov/keras-monotonic-attention#10]
When I run sequential_example.py, I get a warning as :
```
/root/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/gradients_impl.py:112: UserWarning: Converting sparse IndexedSlices to a dense Tensor of unknown shape. This may consume a large amount of memory.
  "Converting sparse IndexedSlices to a dense Tensor of unknown shape. "
```
I try to fix it by using tf.dynnamic_partition instead of tf.gather .
```
import tensorflow as tf

temp = tf.range(0,10)*10 + tf.constant(1,shape=[10])
index = [1,5,9]
temp2 = tf.gather(temp,index)

partitions = tf.reduce_sum(tf.one_hot(index, tf.shape(temp)[0], dtype='int32'), 0)
last_timesteps = tf.dynamic_partition(temp, partitions, 2)
last_timesteps = last_timesteps[1]

with tf.Session() as sess:
    print(sess.run(temp))
    print(sess.run(temp2))
    print(sess.run(last_timesteps))
```
It output:
```
[ 1 11 21 31 41 51 61 71 81 91]
[11 51 91]
[11 51 91]
```
So I try to fix the bug by：
```
# ytm = K.gather(self.E_y, K.cast(ytm, 'int32'))
partitions = tf.reduce_sum(tf.one_hot(K.cast(ytm, 'int32'), tf.shape(self.E_y)[0], dtype='int32'), 0)
ytm = tf.dynamic_partition(self.E_y, partitions, 2)
ytm = ytm[1]
```
but fail.
some useful links：
- https://stackoverflow.com/questions/45882401/how-to-deal-with-userwarning-converting-sparse-indexedslices-to-a-dense-tensor
- https://stackoverflow.com/questions/35892412/tensorflow-dense-gradient-explanation

**Comment by SunYanCN** [comment]
Hope you can fix it!

**Comment by SunYanCN** [comment]
![image](https://user-images.githubusercontent.com/42198591/51649513-ba84f100-1fbf-11e9-982f-bfc0f8caa602.png)
Attention ytm's shape:(?,10).
? is the bug's reason.

**Comment by reyzharkov** [comment]
Yeah, I ignored that issue for a long time. I think more clear and more keras way to solve this problem is to use internal Embedding sublayer. 

Actually now (as this repo is quite old already) there is no need to implement rnn-like behaviour from scratch, it is possible to use another sublayer like [gru cell](
https://github.com/keras-team/keras/blob/master/keras/layers/recurrent.py#L1164) or probably implementing subclass of such cell layer and just use `RNN(cell)` as AttentionDecoder

**Comment by reyzharkov** [comment]
Pushed changes, now it should work without such warning

**Comment by SunYanCN** [comment]
Thanks!
